rosbag2_cpp/src/rosbag2_cpp/reindexer.cpp

```diff
@@ -0,0 +1,320 @@
+
+  // Record the relative paths to the metadata
+  for (const auto & path : files) {
+    auto cleaned_path = strip_parent_path(path);
+    metadata_.relative_file_paths.push_back(cleaned_path);
+  }
+}
+
+void Reindexer::aggregate_metadata(
+  const std::vector<rcpputils::fs::path> & files,
+  const rosbag2_storage::StorageOptions & storage_options)
+{
+  // In order to most accurately reconstruct the metadata, we need to
+  // visit each of the contained relative database files in the bag,
+  // open them, slurp up the info, and stuff it into the master
+  // metadata object.
+  auto metadata_io_default = std::make_unique<rosbag2_storage::MetadataIo>();
+  auto bag_reader = std::make_unique<rosbag2_cpp::readers::SequentialReader>(
+    std::move(storage_factory_), converter_factory_, std::move(metadata_io_default));
+  ROSBAG2_CPP_LOG_INFO_STREAM("Extracting metadata from database(s)");
+  for (const auto & f_ : files) {
+    ROSBAG2_CPP_LOG_INFO_STREAM("Extracting from file: " + f_.string());
+
+    // Set up reader
+    rosbag2_storage::StorageOptions temp_so = {
+      f_.string(),  // uri
+      storage_options.storage_id,  // storage_id
+      storage_options.max_bagfile_size,  // max_bagfile_size
+      storage_options.max_bagfile_duration,  // max_bagfile_duration
+      storage_options.max_cache_size,  // max_cache_size
+      storage_options.storage_config_uri  // storage_config_uri
+    };
+
+    // We aren't actually interested in reading messages, so use a blank converter option
+    rosbag2_cpp::ConverterOptions blank_converter_options {};
+    ROSBAG2_CPP_LOG_INFO_STREAM("Preparing to open...");
+    bag_reader->open(temp_so, blank_converter_options);
+    ROSBAG2_CPP_LOG_INFO_STREAM("Opened");
+    // Get metadata from the bag file
+    auto temp_metadata = bag_reader->get_metadata();
+
+    if (temp_metadata.starting_time < metadata_.starting_time) {
+      metadata_.starting_time = temp_metadata.starting_time;
+    }
+    // ROSBAG2_CPP_LOG_INFO_STREAM(
+    //  "Current duration: " + std::to_string(metadata_.duration.count()));
+    // ROSBAG2_CPP_LOG_INFO_STREAM(
+    //   "Incoming duration: " + std::to_string(temp_metadata.duration.count()));
+    metadata_.duration += temp_metadata.duration;
+    ROSBAG2_CPP_LOG_INFO_STREAM("New duration: " + std::to_string(metadata_.duration.count()));
+    metadata_.message_count += temp_metadata.message_count;
+
+    // Add the topic metadata
+    for (const auto & topic : temp_metadata.topics_with_message_count) {
+      auto found_topic = std::find_if(
+        metadata_.topics_with_message_count.begin(),
+        metadata_.topics_with_message_count.end(),
+        [&topic](const rosbag2_storage::TopicInformation & agg_topic)
+        {return topic.topic_metadata.name == agg_topic.topic_metadata.name;});
+      if (found_topic == metadata_.topics_with_message_count.end()) {
+        // It's a new topic. Add it.
+        metadata_.topics_with_message_count.emplace_back(topic);
```

Comment:
just a thought - you could use an `std::map` for the interim while aggregating data so that you don't have to do as much `find`ing, then convert to a vector at the end

Reply:
Good thinking! Implementing it now...